src/github.com/getlantern/flashlight/geolookup/geolookup.go

```diff
@@ -13,58 +17,92 @@ import (
-	defer lookupMutex.Unlock()
+// Configure configures geolookup to use the given http.Client to perform
+// lookups. geolookup runs in a continuous loop, periodically updating its
+// location and publishing updates to any connected clients. We do this
+// continually in order to detect when the computer's location has changed.
+func Configure(newClient *http.Client) {
+	cfgMutex.Lock()
+	defer cfgMutex.Unlock()
 
-	if lookupData != nil {
-		// We already looked up IP's information.
-		return lookupData, nil
-	}
+	client.Store(newClient)
 
-	var err error
-	for i := 0; i < maxRetries; i++ {
-		// Will look up only if we're using a proxy.
-		lookupData, err = geolookup.LookupIPWithClient("", client)
-		if err == nil {
-			// We got what we wanted, no need to query for it again, let's exit.
-			return lookupData, nil
+	if service == nil {
+		err := registerService()
+		if err != nil {
+			log.Errorf("Unable to register service: %v", err)
+			return
 		}
+		go write()
+		go read()
+		log.Debug("Running")
 	}
-
-	return nil, fmt.Errorf("Unable to look up geolocation information in %d tries: %v", maxRetries, err)
 }
 
-// StartService initializes the geolocation websocket service using the given
-// http.Client to do the lookups
-func StartService(client *http.Client) (err error) {
+func registerService() error {
 	helloFn := func(write func(interface{}) error) error {
-		city, err := getUserGeolocationData(client)
-		if err != nil {
-			return err
+		location := lastKnownLocation.Load()
+		if location == nil {
+			log.Trace("No lastKnownLocation, not sending anything to client")
+			return nil
 		}
-
-		return write(city)
+		log.Trace("Sending last known location to new client")
+		return write(location)
 	}
 
-	if service, err = ui.Register(messageType, nil, helloFn); err != nil {
-		return fmt.Errorf("Unable to register channel: %q", err)
-	}
+	var err error
+	service, err = ui.Register(messageType, nil, helloFn)
+	return err
+}
 
-	go read()
+func write() {
+	retryWaitTime := 100 * time.Millisecond
+	consecutiveFailures := 0
 
-	return nil
+	for {
+		// Wait a random amount of time (to avoid looking too suspicious)
+		// Note - rand was seeded with the startup time in flashlight.go
+		n := rand.Intn(int(publishIntervalVariance))
+		wait := basePublishInterval - publishIntervalVariance/2 + time.Duration(n)
+
+		oldLocation := lastKnownLocation.Load()
+		location, err := geolookup.LookupIPWithClient("", client.Load().(*http.Client))
+		if err == nil {
+			oldLocation := lastKnownLocation.Load()
+			if !reflect.DeepEqual(location, oldLocation) {
+				log.Debugf("Location changed")
+				lastKnownLocation.Store(location)
+			}
+			// Always publish location, even if unchanged
+			service.Out <- location
+		} else {
+			log.Errorf("Unable to get current location: %v", err)
+			// When retrying after a failure, wait a different amount of time
+			retryWait := time.Duration(math.Pow(float64(retryWaitTime), float64(consecutiveFailures)))
```

Comment:
Not sure if we want to use pow here, given that `retryWaitTime` is set to 100, if `consecutiveFailures` is, let's say 5 or 6 that would be a huge amount of hours, also the result could overflow fairly easily, even with small values of `consecutiveFailures`.

At this time, looks like `consecutiveFailures` is never changed (value 0), so `retryWait` is anyways always equal to 1, but in that case the above code is never user and we always end up waiting 1 nanosecond.